Draft of use cases paper. (#32)
j3-fortran/generics: J3-Papers/Generics_Use_Cases.txt

but in some cases these are also expected to drive different sets of
requirements.  Use cases that have redundant requirements are
generally suppressed in this paper, with the exception being when an
otherwise redundant use case might resonate with segments of the user
community or the committee.

The identified categories of use cases are:

    * Geniric Algorithms (2.1)
    * Generic Containers (2.2)
    * Usability, verification, and composition (2.3)

2.1 [ALG] Generic algorithms
----------------------------

2.1.1 Simple use case - SWAP

Perhaps the most common use case of all is the desire to provide a
generic template for swapping the values of two entities.  The
nontrivial aspect of this operation is the need to declare an
auxiliary variable to hold one of the values during processing.

The Fortran generics facility should provide the means to declare a
SWAP template that is applicable for any type that supports copy
assignment.

Notionally, the instantiation of the template for a type T should be
equivalent to:

   SUBROUTINE SWAP(a, b)
      TYPE(T), INTENT(INOUT) :: a
      TYPE(T), INTENT(INOUT) :: b

>       TYPE(T) :: tmp

Author: Good point.   And there is also `TYPEOF` which was in my original draft.  It was suggested that I delete that, so I did.

ASSOCIATE won't work in some of the variants, so I don't think I'll switch to that, but I otherwise would.

still provide powerful, efficient tools within those more specialized
domains.  

Example: Block matrix algebra

Mathematical descriptions of algorithms are often expressed in terms
of block matrices (https://en.wikipedia.org/wiki/Block_matrix).  While
such algorithms can always be "flattened" into algorithms working on
simple matrices, this conversion is tedious and error prone and
generally makes implementations difficult to follow.  Some developers
may find it useful to implement a template that performs matrix
operations on block matrices where types of the blocks are specified
as types that support certain numerical operations.  Note that each
block might itself be a block matrix.

As a concrete example, consider simple matrix-matrix multiplication on 

                C = A x B

Here A, B, and C are m by n, n by p, and m by p arrays of default
reals.  The (unoptimized) multiplication algorithm can be written as

    INTEGER :: i, j, k
    INTEGER :: m, n, p
    REAL :: A(m,n)
    REAL :: B(n,p)
    REAL :: C(m,p)

    DO i = 1, m
       DO j = 1, p
          C(i,j) = 0
          DO k = 1, n
             C(i,j) = C(i,j) + A(i,k)*B(k,j)
	  END DO
       END DO
    END DO

> 

Author: I think that the use of SUM also constrains the types that can be supported in a subtle way.  I doubt it ever really matters.

But in the original version type TC must be addable to the type TA*TB.  In the SUM version   TA*TB must be addable to itself.  (ANd similar twists for the assignment.)

> 

Author: Note that this also does not sidestep the need for a zero-initialization - it merely hides it a bit more.    For instance what if n=0.

	  END DO
       END DO
    END DO

END SUBROTINE GENERIC_MATMUL

The required properties on types TA, TB, and TC are:

    * Type TC must support an operation that returns a generalized "0"

    * Type TA must support (matrix) multiplication by type TB
      returning an intermediate type V

    * Type TC must support addition by type V returning intermediate
      type U

    * Type TC must support assignment to an intermediate type U

Note 1: These intermediate types may be the same as any of TA, TB, and
TC, but the algorithm here is intended to maximize generality.

Note 2: The "scalar" multiplication of types TB and TB in the nested
loop might itself be defined by using the same template, but the
template itself is not recursive.

Note 3: For performance purposes, it is important that the generic
facilities have some mechanism that allows the developer to default
over to optimized libraries (e.g., LAPACK) for types that correspond
to single and double precision reals and complex

Straw votes:

ALG-9: Should the Fortran generics facility provide sufficient
       generality to define numerical algorithms such as the
       generalized linear algebra described above?

>        (YES,NO,UNDECIDED,ABSTAIN)

Author: Interesting.

              to develop templates for generic containers of arbitrary
	      type T? (YES,NO,UNDECIDED,ABSTAIN)

CONTAINERS-2: Should Fortran's generics facilities enable developers to
              create convenient iterators?  (YES,NO,UNDECIDED,ABSTAIN)
	      
CONTAINERS-3: Should Fortran's generics facilities allow developers to
              define easy-to-use accessors both for obtaining objects
              and for assiging values to objects? (YES,NO,UNDECIDED,ABSTAIN)

CONTAINERS-4: Should Fortran's generics facilities provide
              functionality that would allow a well-written container
              to work easily with types that have kind and/or length
              type parameters?  (YES,NO,UNDECIDED,ABSTAIN)

CONTAINERS-5: Should Fortran's generics facilities provide
              functionality that allows a well-written container to
              handle pointers? (YES,NO,UNDECIDED,ABSTAIN)

CONTAINERS-6: Should Fortran's generics facilities enable users of
              relevant containers to indepndently specify a comparison
              operation at instantiation?  (YES,NO,UNDECIDED,ABSTAIN)

> 

Author: I've added it back.

To: J3                                                     J3/##-###
From: T. Clune and generics subgroup
Subject: 
Date: 2021-June-07

References: N2147, 18-110r1

1. Introduction
===============

In WG5/N2147, "Fortran 202X Feature Survey Results -- Final", by Steve
Lionel, the proposed F202X feature that had the most interest and the
most comments was "Generic Programming or Templates". This feature had
a score of 6.21, compared to the next highest score of 5.04 for
"Automatic Allocation on READ into ALLOCATABLE character or array". It
also had almost four pages of comments, while no other feature had as
much as a full page of comments. In spite of the obvious demand for
this feature, its complexity made it inappropriate to include in
Fortran 202X. Instead it was planned that preliminary work on this
feature would be fitted into the schedule of work on Fortran 202X in
the hope that this would allow the completion of the work during the
development of Fortran 202Y.

> Any syntax suggestions in this paper are notional, and only for the

Reviewer: ```suggestion
Any syntax suggestions in this paper are notional, and only for
```

illustrative purposes.

This paper uses the term 'template' to refer to a generic entity that
can be 'instantiated' into concrete implementations using sets of
input 'template parameters'. While there are other possible approaches
to generic programming, (e.g., intelligent macros), the guidance from
J3 in the formation of the generics subgroup was to pursue something
that is at least roughly along the lines of the template approach in
other languages.  We also note that previous work on Parameterized

> Modules easily fits within this intentionally vague definitions of

Reviewer: ```suggestion
Modules easily fits within the intentionally vague definition of
```

templates in this context.

NOTE 1: This paper is intended to be a living document.  As we develop
the F202Y generics features, these use cases will be used for
traceability in the subsequent requirements paper.  If additional use
cases are encountered, or existing ones require modifications, new
versions of this use case paper will be submitted for vote.  And if

> any use previously approved use cases are subsequently shown to be

Reviewer: ```suggestion
any previously approved use cases are subsequently shown to be
```

requirements.  Use cases that have redundant requirements are
generally suppressed in this paper, with the exception being when an
otherwise redundant use case might resonate with segments of the user
community or the committee.

The identified categories of use cases are:

    * Geniric Algorithms (2.1)
    * Generic Containers (2.2)
    * Usability, verification, and composition (2.3)

2.1 [ALG] Generic algorithms
----------------------------

2.1.1 Simple use case - SWAP

Perhaps the most common use case of all is the desire to provide a
generic template for swapping the values of two entities.  The
nontrivial aspect of this operation is the need to declare an
auxiliary variable to hold one of the values during processing.

The Fortran generics facility should provide the means to declare a
SWAP template that is applicable for any type that supports copy
assignment.

Notionally, the instantiation of the template for a type T should be
equivalent to:

   SUBROUTINE SWAP(a, b)
      TYPE(T), INTENT(INOUT) :: a
      TYPE(T), INTENT(INOUT) :: b

      TYPE(T) :: tmp

      tmp = a ! note the use of copy assignment operator '='
      a = b
      b = tmp

   END SUBROUTINE SWAP

> A few variations on SWAP functionality are worth considering. These
> are primary intended as proxies for analogous issues that naturally

Reviewer: ```suggestion
A few variations of SWAP functionality are worth considering. These
are primarily intended as proxies for analogous issues that naturally
```

arise in a wide set of use cases that attempt to be as general as
possible.  In particular these variants emphasize the important
wrinkle that Fortran "type" does not include kind, rank, and other
attributes.

Variant 1: Arbitrary rank

The SWAP instantiation above is designed to work with scalars, but the
underlying algorithm is equally applicable to arrays of arbitrary (but
consistent) shape.  Rank agnostic features planned for F202X are of
significant value in this context, but still require a lengthy SELECT
RANK construct.  It is desirable to provide simple means that enable
developers to declare templates that are rank agnostic.  E.g., one
approach could be to allow integer template parameters that allow the
rank to be specified explicitly and then more easily combined with

> rank already-planned rank-agnostic features.

Reviewer: ```suggestion
already-planned rank-agnostic features.
```

Variant 2: Parameterized types

The instantiation above works for types that lack is intended to work
with types , but the underlying algorithm is equally applicable to
parameterized types provide that the actual type parameters of both
dummy arguments agree.  It is desirable to provide developers with a
simple means to declare templates that work with arbitrary

> parameterized types subject to appropriate requirements.

Reviewer: This paragraph got a little jumbled. I expect it is saying that the new functionality must interoperate with existing parameterized derived types, but I can't quite make sense of it.

that returns a logical when passed two arguments of type T.

We note here that one approach to satisfying this use case might be to
simply relax the requirements on some intrinsics.  While this might
itself be desirable, our intent here is to introduce appropriately
powerful capabilites to allow Fortran developers to define new
templates with comparable functionality.  Intrinsic procedures
themselvs are merely the tip of the proverbial iceberg of relevant use
cases, but serve as examples that are readily familiar to all.

Straw vote:

ALG-5: Should the generics facility enable developers to define
       templates that extend Fortran intrinsics (FINDLOC, MINLOC, etc)
       to work with user defined types which support any operations
       required for a given algorithm? (YES,NO,UNDECIDED,ABSTAIN)

2.1.3 Sorting and Searching
---------------------------

Sorting

While the need to sort a list of items is somewhat rarer in typical
Fortran applications than in wider software communities, it
nonetheless arises often enough to be a problem. It is typically
applied to a rank one array, but sorting algoritms can usually be

> generalized to other categorise of containers with appropriate

Reviewer: ```suggestion
generalized to other categories of containers with appropriate
```

iterators, e.g. Lists and Vectors. A given sorting algorithm can
generally be applied to any type that provides a comparison (< or >)
operation, or equivalent binary logical predicate. With current
Fortran capabilities, the algorithm must be re-implemented for each

> type.  We note however, that some simplification possible through the

Reviewer: ```suggestion
type.  We note however, that some simplification is possible through the
```

> use of include files and newly the newly introduced TYPEOF.

Reviewer: ```suggestion
use of include files and the newly introduced TYPEOF.
```

A generic sorting procedure has the following requirements:

* the ability to have a general type as a generic parameter either as
  an explicit parameter or as implicit in the iterators;

* the ability to associate an assignment operation with a type
  parameter, either implicitly as a property of all types, or as a
  property of the type definition, or as an explicit parameter to the
  generic with a defined interface;

* the ability to associate a comparison operation with a generic type
  parameter, either implicitly as a property of the type definition,
  or as an explicit parameter to the generic with a defined interface;

* and the ability to take one or more iterators as instantiation
  parameters.

It is often necessary to sort the same data according to different
comparison criteria.  It is therefore desirable that the Fortran
generics facility provide means to specify the comparison operation as
a separate template parameter.  This functionality is less crucial

> here than it is for the containers use case above, as in many cases

Reviewer: ```suggestion
here than it is for the containers use case below, as in many cases
```

  generic with a defined interface;

* the ability to associate a comparison operation with a generic type
  parameter, either implicitly as a property of the type definition,
  or as an explicit parameter to the generic with a defined interface;

* and the ability to take two iterators as instantiation parameters.

Straw votes:

ALG-6: Should Fortran generics facilities enable developers to
       implement a sorting template that works with any type
       satisfying the criteria listed above?
       (YES,NO,UNDECIDED,ABSTAIN)

ALG-7: Should Fortran generics facilities enable developers to
       implement a searching template that works with any type satisfying the
       criteria listed above?  (YES,NO,UNDECIDED,ABSTAIN)

ALG-8: Should Fortran generics facilities enable developers to
       implement a merge template that works with any type satisfying the
       criteria listed above?  (YES,NO,UNDECIDED,ABSTAIN)

2.1.4 Complex numerical algorithms
----------------------------------

Many complex numerical algorithms are written for specific intrinsic
types and kinds, but are actually applicable to a broader set of

> types.  The applicability of such algoritms is often narrower than the

Reviewer: ```suggestion
types.  The applicability of such algorithms is often narrower than
```

such algorithms can always be "flattened" into algorithms working on
simple matrices, this conversion is tedious and error prone and
generally makes implementations difficult to follow.  Some developers
may find it useful to implement a template that performs matrix
operations on block matrices where types of the blocks are specified
as types that support certain numerical operations.  Note that each
block might itself be a block matrix.

As a concrete example, consider simple matrix-matrix multiplication on 

                C = A x B

Here A, B, and C are m by n, n by p, and m by p arrays of default
reals.  The (unoptimized) multiplication algorithm can be written as

    INTEGER :: i, j, k
    INTEGER :: m, n, p
    REAL :: A(m,n)
    REAL :: B(n,p)
    REAL :: C(m,p)

    DO i = 1, m
       DO j = 1, p
          C(i,j) = 0
          DO k = 1, n
             C(i,j) = C(i,j) + A(i,k)*B(k,j)
	  END DO
       END DO
    END DO

This algorithm can be generalized to types TA, TB, TC that satisfy
certain properties:

SUBROUTINE GENERIC_MATMUL(A,B,C)
    INTEGER :: i, j, k
    INTEGER :: m, n, p
    TYPE(TA), INTENT(IN) :: A(m,n)
    TYPE(TB), INTENT(IN) :: B(n,p)

>     TYPE(TC), INTENT(IN) :: C(m,p)

Reviewer: ```suggestion
    TYPE(TC), INTENT(OUT) :: C(m,p)
```

    

    DO i = 1, m
       DO j = 1, p

>           C = C%zero()

Reviewer: ```suggestion
          C(i,j) = C(i,j)%zero()
```

          DO k = 1, n
             C(i,j) = C(i,j) + A(i,k)*B(k,j))
	  END DO
       END DO
    END DO

END SUBROTINE GENERIC_MATMUL

The required properties on types TA, TB, and TC are:

    * Type TC must support an operation that returns a generalized "0"

    * Type TA must support (matrix) multiplication by type TB
      returning an intermediate type V

    * Type TC must support addition by type V returning intermediate
      type U

>     * Type TC must support assignment to an intermediate type U

Reviewer: ```suggestion
    * Type TC must support assignment from an intermediate type U
```

Note 1: These intermediate types may be the same as any of TA, TB, and
TC, but the algorithm here is intended to maximize generality.

> Note 2: The "scalar" multiplication of types TB and TB in the nested

Reviewer: ```suggestion
Note 2: The "scalar" multiplication of types TA and TB in the nested
```

to single and double precision reals and complex

Straw votes:

ALG-9: Should the Fortran generics facility provide sufficient
       generality to define numerical algorithms such as the
       generalized linear algebra described above?
       (YES,NO,UNDECIDED,ABSTAIN)

ALG-10: Should the Fortran generics facility provide the means to
        specialize templates for particular types?  E.g., to default
        to use of optimized libraries for instrinsic reals?
        (YES,NO,UNDECIDED,ABSTAIN)

2.2 [CONTAINERS] Generic Containers
-----------------------------------

As scientific models become more complex, an increasing fraction of
the lines of code are infrastructure as-opposed to direct numerical
calculation. Examples of such infrastructure include software layers
that couple independently developed subsystems, frameworks for
managing distributed parallelism, advanced I/O (e.g.,
checkpoint/restart via NetCDF), etc. Generally these infrastructure
layers evolve as an attempt to avoid code duplication as multiple
parts of the system require similar functionality.

Software "containers" are abstractions that enable aggregating groups

> of related entities for convenient and efficient access of hte

Reviewer: ```suggestion
of related entities for convenient and efficient access of the
```

contained items.  Many distinct varieties of software containers have
been developed and used.  By providing canonical and reliable means to
perform common operations, containers can greatly simplify

> implementation of higher level algorithms.  design and implementation
> of complex algorithms.

Reviewer: ```suggestion
the design and implementation
of higher level complex algorithms.
```

contained elements to be of the same dynamic type. Fortran provides
excellent mechanisms for declaring and constructing arrays of any type
(intrinsic and user-defined) as well as for accessing, storing, and
modifying array members (elements) with succinct, tailored syntax:
tuples of indices within parens:

   a(i,j) = x
   x = a(i,j)
   a(i,j) = a(i,j)**2

But arrays are merely one common variety of container, albeit
exceptionally important for matrices and data represented on logically
rectangular domains.  Other commonly used container categories include
List, Vector, Associative Array (aka Map or Dictionary), Set, Stack,
Queue, etc.  Unlike Array containers, these others ar generally more
dynamic - growing in size as necessary when new elements are added to the
container. Here we briefly describe some of the most common varieties
of containers.

We also note that one approach to supporting the use cases in this
section would be to more directly add specific container varieties as
1st order features of Fortran.  However, the wide range of container
varieties compounded with the even wider range of potential internal

> representations and algorithms suggests instead that language should

Reviewer: ```suggestion
representations and algorithms suggests instead that the language should
```

instead provide developers the means to define high-quality container
implementations.  (We envision that an effort analogous to container
templates in STL would readily emerge in the Fortran developer
community.)

2.2.1 Vector containers

In many ways, Vector containers are quite similar to one-dimensional
arrays.  Both provide for efficient random access of contained
elements via a contiguous block of memory, but Vector containers also
provide methods to insert and delete items at the 'back' of the
container.  If the reserved block of memory is insufficient for a
given insertion, a new larger block is automatically allocated and
existing data is transferred to the new storage.  (Block size is
typically doubled in such cases leading to O(n) asymptotic complexity
for inserting n items.)  Removal of items at other locations is
supported, but the operation is inefficient.

The need for vectors is often encountered when the number of items
that match some criterion is not known in advance.  Lacking such
functionality, Fortran codes often implement these scenarios with 2

> sweeps with nearly identical logic.  The first sweeep merely counts

Reviewer: ```suggestion
sweeps with nearly identical logic.  The first sweep merely counts
```

the number of matches and then allocates an array.  The second sweep
then populates the array.  Use of containers makes such algorithms
much simpler to understand in many respects.

> * As with arrays, vector require no operations on the contained type

Reviewer: ```suggestion
* As with arrays, vectors require no operations on the contained type
```

  except assignment.

2.2.2 Set containers

Set containers are intended to store a sorted set of _unique_ objects
and provides logarithmic complexity for searching, removal and
insertion.

* Set containers do not support random access, and therefore
  require some form of iterator for traversing contained items.

* Set containers require that supported types provide an ordering
  operation.

2.2.3 Associative array (aka Map or Dictionary)

> Map containers store elements that are formed by a combinations of
> (unique( keys and values, and thus generally involved 2 distinct
> types.  To some degree, these containers can be regarded of as a

Reviewer: ```suggestion
Map containers store elements that are formed by a combination of
unique keys and values, and thus generally involve 2 distinct
types.  To some degree, these containers can be regarded as a
```

  assignment.

* Some Map containers require that the key type provides an ordering
  operation, while others require that there is a hash function.

2.2.4 Miscellaneous containers

The specific containers mentioned above are among the most common, but
it is worth mentioning some other somewhat common varieties:

* Bitset
* List (singly linked, doubly linked)

2.2.5 Iterators

Iterators provide a simple/consistent mechanism to efficiently loop
through all of the elements in a container, or in some cases subsets
of a container. For containers such as Vector and Array, these are
relatively simple, but for containers such as Set and Map that are
implemented with binary trees or hash buckets, the iterator
abstraction is of exceptional value to the user of the
container. Iterators simultaneously hide complexity and encourage safe
coding styles. In pseudo-code iterator usage is typically something
like:

   < declare container > C
   < declare container iterator > I
   < declare element > e

   I = begin(C)
   loop while I is not end(C)
      e = get(I)
      < do something with e >
     next(I)
   end loop

It is desirable that the generics facility provides features which

> will allow container implementation to provide companion iterator

Reviewer: ```suggestion
will allow container implementations to provide companion iterator
```

concrete example can be found in the block matrix multiplication use
case described in the previous section.

2.3.1 Usability

Developers of such a template desire to have a clear, concise, and
canonical way to articulate requirements on the template type
parameters.  Likewise users of such templates would prefer to
determine in advance whether their own types will satisify the
requirements of a template.  Lacking such a mechanism, users must
resort to trial-and-error.  Further, if their types do not quite work,
they are at the mercy of the quality of compiler error messages to
infer whether any deficiencies in their types might be remedied.

2.3.2 Verification

Developers of complex templates may not only desire to communicate
their template requirements to potential users, but also verify that
instantiations will produce correct code for all types that meet the
requirements.  Here we envision the case that a developer might miss
documenting some necessary operation in the middle of complex code.
The developer tests the template with types that just happen to
support that undocumented operation, but some end user is not so
fortunate.

> At the same time, it is recognized that there is value to having
> enforcement of verification protections might be inconvenient during

Reviewer: ```suggestion
At the same time, while it is recognized that there is value to having
enforcement of verification, these protections might be inconvenient during
```

prototyping and/or introduction of additional scaffolding code.

2.3.3 Composition

Consider a template that in turn uses one or more nontrivial templates
internally.  The requirements on such a template are a superset of the
requirements for the internally used templates.  To avoid tedious
duplication, and the associated risks for long term maintenance, it
should be possible to encapsulate template requirements as named
entities that can be exported and reused.  (Bordering on a requirement
here, but trying to express as a use case.)

Straw votes:

CONCEPTS-1: Should template developers be able to clearly and
            concisely communicate the implied requirements on the
            types used in the template?  (YES,NO,UNDECIDED,ABSTAIN)

CONCEPTS-2: Should template developers be able to verify that their

>             templates do not use any detect unspecified implied

Reviewer: ```suggestion
            templates do not use any unspecified implied
```

            requirements on the types used in the templates?
            (YES,NO,UNDECIDED,ABSTAIN)

CONCEPTS-3: Should it be possible to encapsulate template requirements

>             for reuse in multiple contexts? (YES,NO,UNDECIDED,ABSTAIN)

Reviewer: This doesn't necessarily imply that they would be composable. We should have an additional straw vote on whether they should be composable.

Author: This is the key capability that would enable such composition.    I debated saying something like 'named encapsulation', but thought this should be enough for us to wedge in the desired requirement.

If the encapsulations are named, what more do you think is necessary for composition?

Reviewer: Just being able to name them isn't sufficient. We must explicitly say that it be possible to define new requirements in terms of existing ones. It would be entirely possible to design the semantics and syntax in such a way that enabling that in the future becomes much more difficult.

Author: Naming is a specific mechanism for reuse, and Im trying as much as possible not to imply a particular mechanism.   E.g., maybe the requirements themselves are not named but there is some way to say "and use the requirements from template X parameter Y"  (probably not, but it's just the first thing that popped into my head).

And I don't really think we are talking about defining new requirements in terms of existing ones.  We are talking about _including_ existing requirements that have been defined elsewhere.    Or maybe I'm just splitting hairs.   I'll think some more and come back after I've dealt with easier issues.

Author: How about this wording?   "CONCEPTS-3: Should it be possible to encapsulate template requirements
            for reuse in a composable manner? (YES,NO,UNDECIDED,ABSTAIN)"

Reviewer: I think I'd be happy with that.
